### PR TITLE
[async] Patch StateFlowGraph

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #742

There will probably be caveats but at least this should allow us to do fusion. 

Data flow edges/dependency edges are distinguished via `Node::has_state_flow`.

The function names will be updated after #1859 is merged to avoid merge conflicts. @xumingkuan please feel free to merge this unconditionally for #1859

sfg.pdf

Screenshot from 2020-09-12 01-35-53

war.pdf

Screenshot from 2020-09-12 01-35-45

(For some reason `x[...] = 1` leads to activation, hence the mask)

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/program/state_flow_graph.cpp [hunk 1]
```diff
@@ -35,6 +41,20 @@ void StateFlowGraph::insert_task(const TaskLaunchRecord &rec,
...
+    if (latest_state_readers_.find(output_state) ==
+        latest_state_readers_.end()) {
+      latest_state_readers_[output_state].insert(initial_node_);
+    }
+    for (auto &d : latest_state_readers_[output_state]) {
+      // insert a dependency edge
+      insert_state_flow(d, node.get(), output_state);
+    }
+    latest_state_readers_[output_state].clear();
+  }
```
Comment (k-ye) [inline taichi/program/state_flow_graph.cpp:53]: WAW is still missing..?

```
for os in task_meta.output_states:
  for d in latest_state_owners_[os]:
    insert_state_flow(d, node, os)  # WAW
```

File: taichi/program/state_flow_graph.h [hunk 2]
```diff
@@ -28,14 +28,42 @@ class StateFlowGraph {
...
+      // true: (state flow edge) the state generated by this node is used by the
+      // destination node.
+
+      // false: (dependency edge) the destination node does not
+      // use the generated state, but its execution must happen after this case.
+      // This usually means a write-after-read (WAR) dependency on state.
+
+      // Note:
+      // Read-after-write leads to data flow edges
+      // Write-after-write leads to data flow edges
```
Comment (k-ye) [inline taichi/program/state_flow_graph.h:52]: I feel like WAW is also a dependency, which prevents overly aggressive reordering? I.e. the second write doesn't use the result of the first write, but it must happen after the first write.

File: taichi/program/state_flow_graph.h [hunk 3]
```diff
@@ -28,14 +28,42 @@ class StateFlowGraph {
     std::string task_name;
     // Incremental ID to identify the i-th launch of the task.
     int launch_id;
+    bool is_initial_node{false};
+
+    std::unordered_set<AsyncState, AsyncStateHash> input_states, output_states;
```
Comment (k-ye) [inline taichi/program/state_flow_graph.h:33]: TODO: In the future, maybe we can just store a pointer/reference to `TaskMeta`..

**Review by k-ye — COMMENTED** [review]
Thanks! I only have questions on WAW..

File: taichi/program/async_engine.cpp [hunk 4]
```diff
@@ -264,50 +264,47 @@ TaskMeta AsyncEngine::create_task_meta(const TaskLaunchRecord &t) {
...
         }
       }
     }
     if (auto global_store = stmt->cast<GlobalStoreStmt>()) {
       if (auto ptr = global_store->ptr->cast<GlobalPtrStmt>()) {
         for (auto &snode : ptr->snodes.data) {
-          meta.output_states.emplace_back(snode, AsyncState::Type::value);
-          if (ptr->activate)
-            meta.output_states.emplace_back(snode, AsyncState::Type::mask);
+          meta.input_states.emplace(snode, AsyncState::Type::value);
```
Comment (xumingkuan) [inline taichi/program/async_engine.cpp:274]: Why do we have `input_states` here? It looks like mixing things like RAW and WAW up.

File: taichi/program/state_flow_graph.h (hunk 2 above)
Comment (xumingkuan) [reply taichi/program/state_flow_graph.h:52]: Agreed. WAW and WAR can be eliminated by creating another SNode and write to that...

**Review by xumingkuan — COMMENTED** [review]
Thanks! It looks like WAW is treated like RAW. I would suggest separating them in `create_task_meta`.

File: taichi/program/state_flow_graph.h (hunk 2 above)
Comment (xumingkuan) [reply taichi/program/state_flow_graph.h:52]: ...although we don't do that optimization.

File: taichi/program/async_engine.cpp [hunk 5]
```diff
@@ -264,50 +264,47 @@ TaskMeta AsyncEngine::create_task_meta(const TaskLaunchRecord &t) {
     if (auto global_load = stmt->cast<GlobalLoadStmt>()) {
       if (auto ptr = global_load->ptr->cast<GlobalPtrStmt>()) {
         for (auto &snode : ptr->snodes.data) {
-          meta.input_states.emplace_back(snode, AsyncState::Type::value);
+          meta.input_states.insert({snode, AsyncState::Type::value});
```
Comment (k-ye) [inline taichi/program/async_engine.cpp:267]: nit: for consistency..

```suggestion
          meta.input_states.emplace(snode, AsyncState::Type::value);
```

**Review by k-ye — APPROVED** [review]
LGTM!

**Review by xumingkuan — APPROVED** [review]
LGTM now!

**Pull request merged**